### PR TITLE
Stats table

### PR DESCRIPTION
**Pull request by Sciguymjm**

### CODE REVIEW

File: backend/database/wrapper/stats/player_stat_wrapper.py [hunk 1]
```diff
@@ -36,6 +36,13 @@ def __init__(self, player_wrapper: PlayerWrapper):
 
     def get_stats(self, session, id_, stats_query, std_query, rank=None, redis=None, raw=False, replay_ids=None,
                   playlist=13, win: bool = None):
+
+
+        def float_maybe(f):
```
Comment (dtracers) [inline backend/database/wrapper/stats/player_stat_wrapper.py]: can we make this a util method?
This does seem to have a need be nested

File: backend/utils/psyonix_api_handler.py [hunk 2]
```diff
@@ -165,14 +165,14 @@ def get_empty_data(ids):
...
-        for mode in modes.values()
-    ]
+    return {id:
+                {'mode': mode,
+                 'rank_points': rank * random.randint(60, 75) + 15 * div,
+                 'tier': rank,
+                 'division': div,
+                 'string': tier_div_to_string(rank, div)}
+            for id, mode in modes.items()
+            }
```
Comment (dtracers) [inline backend/utils/psyonix_api_handler.py]: this does not match brackets

File: webapp/src/Requests/Player/getPlayStyle.ts [hunk 3]
```diff
@@ -42,3 +42,316 @@ export const getPlayStyle = (id: string, rank?: number, playlist?: number): Prom
     const url = qs.stringify({rank, playlist}, {addQueryPrefix: true, indices: false})
     return doGet(`/player/${id}/play_style` + url)
 }
+
+export const getPlayStyleRaw = (id: string, playlist?: number): Promise<PlayStyleRawResponse> => {
```
Comment (dtracers) [inline webapp/src/Requests/Player/getPlayStyle.ts]: can this be in a different file?
Like a mock data file

Comment (dtracers) [reply webapp/src/Requests/Player/getPlayStyle.ts]: the mock data part not the method itself

File: backend/database/wrapper/stats/shared_stats_wrapper.py [hunk 4]
```diff
@@ -66,3 +66,9 @@ def get_wrapped_stats(self, stats, creation: SharedStatCreation):
 
     def get_player_stat_std_query(self):
         return self.player_stats.std_query
+
+    def float_maybe(self, f):
```
Comment (twobackfromtheend) [inline backend/database/wrapper/stats/shared_stats_wrapper.py]: static?

File: backend/utils/psyonix_api_handler.py [hunk 5]
```diff
@@ -165,14 +165,15 @@ def get_empty_data(ids):
...
-    return [
-        {'mode': mode,
-         'rank_points': rank * random.randint(60, 75) + 15 * div,
-         'tier': rank,
-         'division': div,
-         'string': tier_div_to_string(rank, div)}
-        for mode in modes.values()
-    ]
+    return {
+        id:
```
Comment (twobackfromtheend) [inline backend/utils/psyonix_api_handler.py]:
```suggestion
        id_:
```

File: backend/utils/psyonix_api_handler.py [hunk 6]
```diff
@@ -165,14 +165,15 @@ def get_empty_data(ids):
...
+            {'mode': mode,
+             'rank_points': rank * random.randint(60, 75) + 15 * div,
+             'tier': rank,
+             'division': div,
+             'string': tier_div_to_string(rank, div)}
+        for id, mode in modes.items()
```
Comment (twobackfromtheend) [inline backend/utils/psyonix_api_handler.py]:
```suggestion
        for id_, mode in modes.items()
```

File: webapp/src/Components/Player/Compare/PlayStyle/PlayerComparePlayStyleCharts.tsx [hunk 7]
```diff
@@ -37,6 +40,7 @@ export class PlayerComparePlayStyleCharts extends React.PureComponent<Props, Sta
                         indicesToRemove.push(i)
                     }
                 })
+            console.log(indicesToRemove)
```
Comment (twobackfromtheend) [inline webapp/src/Components/Player/Compare/PlayStyle/PlayerComparePlayStyleCharts.tsx]: remove.

File: webapp/src/Components/Player/Compare/PlayStyle/PlayerComparePlayStyleCharts.tsx [hunk 8]
```diff
@@ -57,7 +61,13 @@ export class PlayerComparePlayStyleCharts extends React.PureComponent<Props, Sta
                 .map((playStyleResponse) => playStyleResponse.chartDatas[i])
             )
         const chartTitles = playerPlayStyles[0].chartDatas.map((chartData) => chartData.title)
-
+        const checkbox = (
+            <FormControlLabel
+                control={<Checkbox
```
Comment (twobackfromtheend) [inline webapp/src/Components/Player/Compare/PlayStyle/PlayerComparePlayStyleCharts.tsx]: Checkbox would probably look better on a single line

Comment (twobackfromtheend) [reply webapp/src/Components/Player/Compare/PlayStyle/PlayerComparePlayStyleCharts.tsx]: Also maybe use a switch?

File: webapp/src/Components/Player/Compare/PlayStyle/PlayerComparePlayStyleCharts.tsx [hunk 9]
```diff
@@ -79,6 +89,16 @@ export class PlayerComparePlayStyleCharts extends React.PureComponent<Props, Sta
                         </Grid>
                     )
                 })}
+                <Grid container xs={12} justify="center" alignItems="center">
+                    <Grid item xs={12} style={{textAlign: "center"}}>
+                        {checkbox}
+                    </Grid>
+                    <Grid item xs={"auto"}>
```
Comment (twobackfromtheend) [inline webapp/src/Components/Player/Compare/PlayStyle/PlayerComparePlayStyleCharts.tsx]: string props don't need to be in {}
```suggestion
                    <Grid item xs="auto">
```

File: webapp/src/Components/Player/Compare/PlayStyle/PlayerComparePlayStyleCharts.tsx [hunk 10]
```diff
@@ -79,6 +89,16 @@ export class PlayerComparePlayStyleCharts extends React.PureComponent<Props, Sta
...
+                        <PlayerCompareTable names={players.map((player) => player.name)}
+                                            rawPlayers={playerPlayStylesRaw}
+                        heatmap={this.state.heatmapMode}/>
```
Comment (twobackfromtheend) [inline webapp/src/Components/Player/Compare/PlayStyle/PlayerComparePlayStyleCharts.tsx]: what is this alignment lol

File: webapp/src/Components/Player/Compare/PlayStyle/PlayerCompareTable.tsx [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+
+    public render() {
+        const {names, rawPlayers, heatmap} = this.props
+        const header = (
+            <TableRow>
+                <TableCell>Stat</TableCell>
+                {names.map((player) =>
+                    <>
+                        <TableCell>{player}</TableCell>
+
```
Comment (twobackfromtheend) [inline webapp/src/Components/Player/Compare/PlayStyle/PlayerCompareTable.tsx]: unnecessary blank line, I think the `<> </>` is unnecessary too.

File: webapp/src/Components/Player/Compare/PlayStyle/PlayerCompareTable.tsx [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+            })
+            return (
+                <Table>
+                    <TableHead>
+                        {header}
+                    </TableHead>
+                    <TableBody>
+
+                        {stats.map((stat) => {
+                                return <TableRow key={stat.name}>
```
Comment (twobackfromtheend) [inline webapp/src/Components/Player/Compare/PlayStyle/PlayerCompareTable.tsx]: can replace `=> {return <Thing/>}` with `=> (<Thing/>)` which'll look better i think.

Also remove the empty line above the map.

File: webapp/src/Components/Player/Compare/PlayStyle/PlayerCompareTable.tsx [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+                                    <TableCell>
+                                        {stat.name}
+                                    </TableCell>
+                                    {stat.values.map((value, i) =>
+                                        <TableCell key={i}
+                                                   style={{
+                                                       backgroundColor: ((heatmap && rawPlayers.length > 1) ?
```
Comment (twobackfromtheend) [inline webapp/src/Components/Player/Compare/PlayStyle/PlayerCompareTable.tsx]: extract to a method please.

Comment (twobackfromtheend) [reply webapp/src/Components/Player/Compare/PlayStyle/PlayerCompareTable.tsx]: getTableCellStyles?

File: webapp/src/Components/Player/Compare/PlayStyle/PlayerCompareTable.tsx [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+                                                           `rgba(${200 * (value - stat.min) / (stat.max - stat.min)}, 0,
+                                                            ${200 * (stat.max - value) / (stat.max - stat.min)}, 1)`
+                                                           :
+                                                           "#fff"),
+                                                       color: ((heatmap && rawPlayers.length > 1) ? "white" : "black")
+                                                   }}>
+                                            {value.toFixed(2)}
```
Comment (twobackfromtheend) [inline webapp/src/Components/Player/Compare/PlayStyle/PlayerCompareTable.tsx]: Why not use the roundToDp (not exact name) function in stringUtils. 

That will avoid showing 2 as 2.00

File: webapp/src/Components/Player/Compare/PlayStyle/PlayerCompareTable.tsx [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+                                        </TableCell>)}
+                                </TableRow>
+                            }
+                        )}
+                    </TableBody>
+                </Table>
+            )
+        } else {
+            return ""
```
Comment (twobackfromtheend) [inline webapp/src/Components/Player/Compare/PlayStyle/PlayerCompareTable.tsx]: don't need the else. also should return null?

```suggestion
            return null
```

File: webapp/src/Requests/Player/getPlayStyle.ts [hunk 16]
```diff
@@ -1,44 +1,20 @@
...
     }
     const url = qs.stringify({rank, playlist}, {addQueryPrefix: true, indices: false})
     return doGet(`/player/${id}/play_style` + url)
 }
+
+export const getPlayStyleRaw = (id: string, playlist?: number): Promise<PlayStyleRawResponse> => {
+    if (useMockData) {
+        return Promise.resolve(MOCK_PLAY_STYLE_RAW)
+    }
+    const url = qs.stringify({playlist}, {addQueryPrefix: true, indices: false})
```
Comment (twobackfromtheend) [inline webapp/src/Requests/Player/getPlayStyle.ts]: Variable should probably be better named as queryParams.

**Pull request merged**